File: frontend/src/app/core/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { environment } from '../../../environments/environment';
import { AuthResponse } from '../../models/auth/auth-response';
import { LoginRequest } from '../../models/auth/login-request';
import { RegisterRequest } from '../../models/auth/register-request';
import { TokenService } from './token.service';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private apiUrl = environment.apiUrl;

  constructor(
    private http: HttpClient,
    private tokenService: TokenService
  ) {}

  login(credentials: LoginRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.apiUrl}/auth/login`, credentials)
      .pipe(
        tap(response => {
          this.tokenService.setToken(response.token);
          this.tokenService.setUser(response.user);
        })
      );
  }

  register(data: RegisterRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.apiUrl}/auth/register`, data)
      .pipe(
        tap(response => {
          this.tokenService.setToken(response.token);
          this.tokenService.setUser(response.user);
        })
      );
  }

  /** Alta de usuario sin reemplazar la sesión actual (panel de administración). */
  registerSystemUser(data: RegisterRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.apiUrl}/auth/register`, data);
  }

  logout(): void {
    this.tokenService.removeToken();
  }

  isLoggedIn(): boolean {
    return this.tokenService.isLoggedIn();
  }

  getCurrentUser() {
    return this.tokenService.getUser();
  }
}

File: app_mantenimiento/lib/services/auth_service.dart
import '../config/api_config.dart';
import '../models/auth/auth_response.dart';
import '../models/auth/login_request.dart';
import '../models/auth/register_request.dart';
import 'api_service.dart';

class AuthService {
  Future<AuthResponse> register(RegisterRequest request) async {
    final data = ApiPayload.decodeMap(await ApiService.post(
      ApiConfig.authRegister,
      body: request.toJson(),
      auth: false,
    ));
    final authResponse = AuthResponse.fromJson(data);
    await ApiService.saveToken(authResponse.token);
    return authResponse;
  }

  Future<AuthResponse> login(LoginRequest request) async {
    final data = ApiPayload.decodeMap(await ApiService.post(
      ApiConfig.authLogin,
      body: request.toJson(),
      auth: false,
    ));
    final authResponse = AuthResponse.fromJson(data);
    await ApiService.saveToken(authResponse.token);
    return authResponse;
  }

  Future<AuthResponse> refreshToken() async {
    final data =
        ApiPayload.decodeMap(await ApiService.post(ApiConfig.authRefresh));
    final authResponse = AuthResponse.fromJson(data);
    await ApiService.saveToken(authResponse.token);
    return authResponse;
  }

  Future<void> logout() async {
    await ApiService.removeToken();
  }

  Future<bool> isLoggedIn() async {
    final token = await ApiService.getToken();
    return token != null && token.isNotEmpty;
  }
}
